types/encoding.go

```diff
@@ -19,7 +19,17 @@ func (b Block) MarshalSia(w io.Writer) error {
 	w.Write(b.ParentID[:])
 	w.Write(b.Nonce[:])
 	encoding.WriteUint64(w, uint64(b.Timestamp))
-	return encoding.NewEncoder(w).EncodeAll(b.MinerPayouts, b.Transactions)
+	encoding.WriteInt(w, len(b.MinerPayouts))
+	for i := range b.MinerPayouts {
+		b.MinerPayouts[i].MarshalSia(w)
+	}
+	encoding.WriteInt(w, len(b.Transactions))
+	for i := range b.Transactions {
+		if err := b.Transactions[i].MarshalSia(w); err != nil {
+			return err
+		}
+	}
+	return nil
```

Comment:
Can define a MarshalSiaOld function that has the identical implementation to the previous implementation of MarshalSia. Then you can call it and have it pass back a writer. If !DEBUG, all it does is pass back the same writer you sent it, and nothing more. If DEBUG, it'll implement a writer that does a sanity check on the byte stream.

Something like that. Again, we've caught enough bugs with these, and there's enough code change in this PR (covered fields as well as block as well as sighash) that I think we really need a sanity check.